Update payload.mdx (#2821)
withastro/docs: src/content/docs/en/guides/cms/payload.mdx

service: Payload CMS
---

[PayloadCMS](https://payloadcms.com/) is a headless open-source content management system that can be used to provide content for your Astro project.

## Integrating with Astro

## Prerequisites

To get started, you will need to have the following:

1. **An Astro project** - If you don't have an Astro project yet, our [Installation guide](/en/install/auto/) will get you up and running in no time.
2. **A MongoDB database** - PayloadCMS will ask you for a MongoDB connection string when creating a new project. You can set one up locally or use [MongoDBAtlas](https://www.mongodb.com/) to host a database on the web for free.
3. **A PayloadCMS REST API** - Create a [PayloadCMS](https://payloadcms.com/) project and connect it to your MongoDB database during the setup.

### Installing PayloadCMS

> Follow the [PayloadCMS Installation docs](https://payloadcms.com/docs/getting-started/installation) in order to install a fresh PayloadCMS app. During the PayloadCMS installation, you will be asked if you want to use a template.
> 
> If you pick one of the available templates (such as 'blog'), additional collections will be generated for you. You can then skip the following section, because you already have a collection other than "Users" that you can use.
> 
> If you don't pick any templates, only a single collection called "Users" will be generated.
> 
> If you start the PayloadCMS server using `npm run dev`, you can go to `http://localhost:3000/admin` to enter your PayloadCMS Dashboard for the first time. Create the initial administrator account and enter the Dashboard. You should now see that the "Users" collection contains just the administrator account.

Reviewer: ```suggestion
1. Follow the [PayloadCMS Installation docs](https://payloadcms.com/docs/getting-started/installation) in order to install a fresh PayloadCMS app. 

:::note[Choosing a template]
During the PayloadCMS installation, you will be asked if you want to use a template.

If you do not choose a template, only a single collection called "Users" will be generated and you will have to [define a PayloadCMS collection manually](#creating-a-payload-cms-collection) in Astro to use for your content.

Choosing any of the available templates at this step (such as 'blog') automatically generates additional collections for you to use.
:::

2. Start the PayloadCMS server using `npm run dev` and visit `http://localhost:3000/admin` to enter your PayloadCMS Dashboard for the first time. 

3. Using the Dashboard, create an initial administrator account, which should now be visible as the only entry in your "Users" collection.
```

Reviewer 2: one thing to note here is that the cms will not start if a valid mongo connection is not established. Maybe it will be a good thing to expand a bit on how to create a new mongo instance locally or using Atlas since this blocks the following steps

Author: > one thing to note here is that the cms will not start if a valid mongo connection is not established. Maybe it will be a good thing to expand a bit on how to create a new mongo instance locally or using Atlas since this blocks the following steps

This is explained in the PayloadCMS docs. I can expand on it here as well if needed, of course.

#### Creating a PayloadCMS Collection

> You can create a new Collection called `Posts.ts` in `src/collections`.

Reviewer: ```suggestion
If you did not choose a template during installation that created a content collection for you, create a new Payload CMS Collection by adding a configuration file in `src/collections/` called `Posts.ts`.
```

Reviewer 2: I think this section should be more inclined to how to create your own collection and how to make connections between them

> ```astro title="src/collections"

Reviewer: ````suggestion
```astro title="src/collections/Posts.ts"
````

    useAsTitle: "title",
  },
  access: {
    read: () => true,
  },

  fields: [
    {
      name: "title",
      type: "text",
      required: true,
    },
    {
      name: "content",
      type: "text",
      required: true,
    },
    {
      name: "slug",
      type: "text",
      required: true,
    },
  ],
};

export default Posts;
```

This creates a "Posts" collection which requires the following : `title`, `content`, `slug`. You will use this collection to manage your blog posts. The colllection is publicly available, so you shouldn't put any sensitive info inside.

You then need to import and add it to the available collections in the `payload.config.ts` file.

```astro title="src/payload.config.ts"
import { buildConfig } from "payload/config";
import path from "path";

import Users from "./collections/Users";
import Posts from "./collections/Posts";

export default buildConfig({
  serverURL: "http://localhost:3000",
  admin: {
    user: Users.slug,
  },
  collections: [Users, Posts],
  typescript: {
    outputFile: path.resolve(__dirname, "payload-types.ts"),
  },
  graphQL: {
    schemaOutputFile: path.resolve(__dirname, "generated-schema.graphql"),
  },
});
```

> If you then go into the PayloadCMS Dashboard, you should see a new collection called "Posts" appear next to the "Users" collection. Enter the "Posts" collection and create a new post. After saving it, you will notice the API URL appear in the bottom right corner.

Reviewer: Just pointing out that I'm not sure where the "skippable steps" stop here. If someone chose a template (e.g. blog) would n original post have been created, so this step of creating a new post should also be skipped?

It also seems like the note about both services using port 3000 should not be in the "skipped" section... so I think *somewhere* here we might need a new subheading so it's more clear where a reader with a template should jump back in to the instructions?

(Note: I did make an earlier suggested edit that took out the instruction to "skip ahead", but then I started the next instruction with "If you didn't choose a template..." So the idea is still the same: we need to make it clear when the "skip/if you choose a template" steps stop so that people are clear about where to jump back in.)

Reviewer 2: Excellent point here, most of these steps can be skipped if the blog template is selected. Personally, I used the UI to create new content

Author: >  Personally, I used the UI to create new content

I think we are talking about the same thing. Do you mean the PayloadCMS Dashboard when you say "UI"?

Reviewer 2: Sorry for the confusion, yes I am talking about the Dashboard UI

> If not already running, start the server using `npm run dev` again.
> 
> Open `http://localhost:3000/api/posts` in your browser. You should see a JSON file containing the post you have created as an object.

Reviewer: ```suggestion
With the dev server running, open `http://localhost:3000/api/posts` in your browser. You should see a JSON file containing the post you have created as an object.
```

Reviewer 2: One thing that I noticed about the rest endpoint is going to need its own type parser to convert objects to html nodes

:::tip
By default, both Astro and PayloadCMS will use the port 3000. You might want to change the PayloadCMS port in the `src/server.ts` file. Don't forget to update the `serverURL` in `src/payload.config.ts` as well.
:::

> ```astro title="http://localhost:3000/api/posts"

Reviewer: Note to Team Docs: 

Have we ever used a *website URL* as a code snippet title before?  I'm not sure what our precedent here is, when you want to tell the reader to navigate to a URL but then what they should see is a code snippet!  :smile:

What do we think? Do we need anything here?

{
   "docs":[
      {
         "id":"64098b16483b0f06a7e20ed4",
         "title":"Astro & PayloadCMS Title 🚀",
         "content":"Astro & PayloadCMS Content",
         "slug":"astro-payloadcms-slug",
         "createdAt":"2023-03-09T07:30:30.837Z",
         "updatedAt":"2023-03-09T07:30:30.837Z"
      }
   ],
   "totalDocs":1,
   "limit":10,
   "totalPages":1,
   "page":1,
   "pagingCounter":1,
   "hasPrevPage":false,
   "hasNextPage":false,
   "prevPage":null,
   "nextPage":null
}
```

### Fetching Data

> Fetch your PayloadCMS data through your site's unique REST API URL and the route for your content. (By default, PayloadCMS will mount all routes through `/api`.) Then, you can render your data properties using Astro's `set:html=""` directive.

Reviewer: Have we mentioned what the site's unique REST API URL is, or where to find it? Is this just depending on the port used? `localhost:portnumber/api/posts`?

Together with your post, PayloadCMS will return some top-level metadata. The actual documents are nested within the `docs` array.

For example, to display a list of post titles and their content:

```astro title="src/pages/index.astro"
---
import HomeLayout from "../layouts/HomeLayout.astro";

> const res = await fetch("http://localhost:5000/api/posts") // https://localhost:3000/api/posts by default

Reviewer: one of these is `https` and the other is `http` and I'm assuming they should be the same?

const posts = await res.json()
---

<HomeLayout title='Astro Blog'>
	{
    posts.docs.map((post) => (
        <h2 set:html={post.title} />
        <p set:html={post.content} />
    ))
	}
</HomeLayout>
```

## Building a blog with PayloadCMS and Astro

> The page `src/pages/index.astro` lists each of your posts with a link to its own page.

Reviewer: ```suggestion
Create a blog index page `src/pages/index.astro` to list each of your posts with a link to its own page.
```

> Fetching via the API an array of objects (posts) that include, among others, the following properties:

Reviewer: ```suggestion
Fetching via the API returns an array of objects (posts) that include, among others, the following properties:
```

- `title`
- `content`
- `slug`

```astro title="src/pages/index.astro"
---
import HomeLayout from "../layouts/HomeLayout.astro";

const res = await fetch("http://localhost:5000/api/posts") // https://localhost:3000/api/posts by default
const posts = await res.json()
---

<HomeLayout title='Astro Blog'>
	<h1>Astro + PayloadCMS 🚀</h1>
	<h2>Blog posts list:</h2>
	<ul>
		{
			posts.docs.map((post) =>(
				<li>
					<a href={`posts/${post.slug}`} set:html={post.title} />
				</li>
			))
		}

	</ul>
</HomeLayout>
```

### Using the PayloadCMS API to generate pages

> The page `src/pages/posts/[slug].astro` [dynamically generates a page](/en/core-concepts/routing/#dynamic-routes) for each post.

Reviewer: ```suggestion
Create a page `src/pages/posts/[slug].astro` to [dynamically generate a page](/en/core-concepts/routing/#dynamic-routes) for each post.
```

i18nReady: false
---

[PayloadCMS](https://payloadcms.com/) is a headless open-source content management system that can be used to provide content for your Astro project.

## Integrating with Astro

> ## Prerequisites

Reviewer 3: Are these supposed to be on the same H level?

Author: Nope, my bad! Fixed.

  admin: {
    user: Users.slug,
  },
  collections: [Users, Posts],
  typescript: {
    outputFile: path.resolve(__dirname, "payload-types.ts"),
  },
  graphQL: {
    schemaOutputFile: path.resolve(__dirname, "generated-schema.graphql"),
  },
});
```

This will make a new collection called "Posts" appear in your PayloadCMS Dashboard next to the "Users" collection. 

2. Enter the "Posts" collection and create a new post. After saving it, you will notice the API URL appear in the bottom right corner.

3. With the dev server running, open `http://localhost:3000/api/posts` in your browser. You should see a JSON file containing the post you have created as an object.

```json
{
   "docs":[
      {
         "id":"64098b16483b0f06a7e20ed4",
         "title":"Astro & PayloadCMS Title 🚀",
         "content":"Astro & PayloadCMS Content",
         "slug":"astro-payloadcms-slug",
         "createdAt":"2023-03-09T07:30:30.837Z",
         "updatedAt":"2023-03-09T07:30:30.837Z"
      }
   ],
   "totalDocs":1,
   "limit":10,
   "totalPages":1,
   "page":1,
   "pagingCounter":1,
   "hasPrevPage":false,
   "hasNextPage":false,
   "prevPage":null,
   "nextPage":null
}
```

> :::tip
> By default, both Astro and PayloadCMS will use port 3000. You might want to change the PayloadCMS port in the `src/server.ts` file. Don't forget to update the `serverURL` in `src/payload.config.ts` as well.
> :::

Reviewer 3: ```suggestion

:::tip
By default, both Astro and PayloadCMS will use port 3000. You might want to change the PayloadCMS port in the `src/server.ts` file. Don't forget to update the `serverURL` in `src/payload.config.ts` as well.
:::
```

i18nReady: false
---

[PayloadCMS](https://payloadcms.com/) is a headless open-source content management system that can be used to provide content for your Astro project.

## Integrating with Astro

## Prerequisites

To get started, you will need to have the following:

1. **An Astro project** - If you don't have an Astro project yet, our [Installation guide](/en/install/auto/) will get you up and running in no time.
2. **A MongoDB database** - PayloadCMS will ask you for a MongoDB connection string when creating a new project. You can set one up locally or use [MongoDBAtlas](https://www.mongodb.com/) to host a database on the web for free.
3. **A PayloadCMS REST API** - Create a [PayloadCMS](https://payloadcms.com/) project and connect it to your MongoDB database during the setup.

> 
> 

Reviewer: ```suggestion
```

i18nReady: false
---

[PayloadCMS](https://payloadcms.com/) is a headless open-source content management system that can be used to provide content for your Astro project.

## Integrating with Astro

## Prerequisites

To get started, you will need to have the following:

1. **An Astro project** - If you don't have an Astro project yet, our [Installation guide](/en/install/auto/) will get you up and running in no time.
2. **A MongoDB database** - PayloadCMS will ask you for a MongoDB connection string when creating a new project. You can set one up locally or use [MongoDBAtlas](https://www.mongodb.com/) to host a database on the web for free.
3. **A PayloadCMS REST API** - Create a [PayloadCMS](https://payloadcms.com/) project and connect it to your MongoDB database during the setup.

> ### Installing PayloadCMS

Reviewer: ```suggestion
```
So I've been thinking about this, and since we already said a Payload project was a prerequisite, maybe the note just directly follows that. And everything necessary to connect Astro is what should be the next heading here?

> 1. Follow the [PayloadCMS Installation docs](https://payloadcms.com/docs/getting-started/installation) in order to install a fresh PayloadCMS app. 

Reviewer: ```suggestion
```

:::note[Choosing a template]
During the PayloadCMS installation, you will be asked if you want to use a template.

> If you do not choose a template, only a single collection called "Users" will be generated and you will have to [define a PayloadCMS collection manually](#creating-a-payload-cms-collection) in Astro to use for your content.

Reviewer: ```suggestion
If you do not choose a template, only a single collection called "Users" will be generated and you will have to [define a PayloadCMS collection manually](#creating-a-payloadcms-collection) in Astro to use for your content.
```

Choosing any of the available templates at this step (such as 'blog') automatically generates additional collections for you to use.
:::

2. Start the PayloadCMS server using `npm run dev` and visit `http://localhost:3000/admin` to enter your PayloadCMS Dashboard for the first time. 

3. Using the Dashboard, create an initial administrator account, which should now be visible as the only entry in your "Users" collection.

> #### Creating a PayloadCMS Collection

Reviewer: ```suggestion
### Configuring Astro for your  PayloadCMS collection
```
So now this can be more like:

> Your Payload project template will contain a file called Posts.ts in src/collections/. The example below shows this file for a collection that requires "title"....

and you can skip to that paragraph
